woocommerce/woocommerce-ios
# Add ManagedObjectModelsInventory

Refs #2473. 

I was trying to analyze how the migration works so I can start writing unit tests before we tackle #2473. I found this logic in CoreDataManager:

https://github.com/woocommerce/woocommerce-ios/blob/e81fdc73468c2ed43a4ec7a3cfc4b6b3df888cf6/Storage/Storage/CoreData/CoreDataManager.swift#L187-L199

It's a bit complex. It's one of the most important parts of the migration since it relies on [our **convention** of creating model versions in sequence](https://github.com/woocommerce/woocommerce-ios/tree/develop/Storage/Storage/Model/WooCommerce.xcdatamodeld) in order to work. 

But it's lacking some unit tests.  

## Changes 

### Extract Models Introspection to ManagedObjectModelsInventory

I originally just wanted to add tests to `CoreDataManager` for that part but it turned out to be pretty difficult. I decided to extract that logic into a new `struct` and factory instead:

https://github.com/woocommerce/woocommerce-ios/blob/ecf9f7f55d15343f2ce9993876146102f0a74e4e/Storage/Storage/CoreData/ManagedObjectModelsInventory.swift#L61-L82

Every step should have some documentation to help us understand the expectations and what's happening. With this extracted logic, it was easier to make [isolated unit tests](https://github.com/woocommerce/woocommerce-ios/blob/issue/2371-add-model-finder-tests/Storage/StorageTests/CoreData/ManagedObjectModelsInventoryTests.swift).

### Use ManagedObjectModelsInventory in CoreDataManager and IterativeMigrator

The new `ManagedObjectModelsInventory` is now used by both [`CoreDataManager`](https://github.com/woocommerce/woocommerce-ios/blob/ecf9f7f55d15343f2ce9993876146102f0a74e4e/Storage/Storage/CoreData/CoreDataManager.swift#L15) and [`CoreDataIterativeMigrator`](https://github.com/woocommerce/woocommerce-ios/blob/ecf9f7f55d15343f2ce9993876146102f0a74e4e/Storage/Storage/CoreData/CoreDataIterativeMigrator.swift#L11). This simplifies CoreDataManager even more. It is now more focused on being the "database stack" instead of having migration-related logic. 

I was planning to extract [this logic](https://github.com/woocommerce/woocommerce-ios/blob/ecf9f7f55d15343f2ce9993876146102f0a74e4e/Storage/Storage/CoreData/CoreDataIterativeMigrator.swift#L59-L88) out of `CoreDataIterativeMigrator`. It seems like a good candidate for `ManagedObjectModelsInventory` or a `UseCase` but this pull request was getting huge. 

This array was **deleted** as a bonus. 🎉 

https://github.com/woocommerce/woocommerce-ios/blob/e81fdc73468c2ed43a4ec7a3cfc4b6b3df888cf6/Storage/StorageTests/CoreData/CoreDataIterativeMigratorTests.swift#L7-L9

## Testing

I think [this test](https://github.com/woocommerce/woocommerce-ios/blob/e81fdc73468c2ed43a4ec7a3cfc4b6b3df888cf6/Storage/StorageTests/CoreData/CoreDataIterativeMigratorTests.swift#L183) covers the iterative migration well so please do a simple migration test to check for regression:

1. `git checkout 3.9`
2. Run and log in.
3. `git checkout -t origin/issue/2371-add-model-finder-tests` (this PR's branch)
4. Run. This should perform a migration from Model 26 to 28.

## Reviewing 

Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist
